### ISSUE TITLE
syntax error in test file results in no tests being run and "PASSED"

### ISSUE THREAD

**Issue by boegel** [eth-cscs/reframe#2302]
```
eessi@ip-172-31-18-112 ~/compatibility-layer $ reframe -r -v -c ./test/compat_layer.py
[ReFrame Setup]
  version:           3.9.1
  command:           '/cvmfs/pilot.eessi-hpc.org/2021.12/compat/linux/x86_64/usr/lib/python-exec/python3.9/reframe -r -v -c ./test/compat_layer.py'
  launched by:       ec2-user@ip-172-31-18-112.eu-west-1.compute.internal
  working directory: '/home/ec2-user/compatibility-layer'
  settings file:     '<builtin>'
  check search path: '/home/ec2-user/compatibility-layer/test/compat_layer.py'
  stage directory:   '/home/ec2-user/compatibility-layer/stage'
  output directory:  '/home/ec2-user/compatibility-layer/output'

/cvmfs/pilot.eessi-hpc.org/2021.12/compat/linux/x86_64/usr/lib/python-exec/python3.9/reframe: skipping test file '/home/ec2-user/compatibility-layer/test/compat_layer.py': reframe syntax error: accessing a test parameter from the class body is disallowed (rerun with '-v' for more information)
Traceback (most recent call last):
  File "/cvmfs/pilot.eessi-hpc.org/2021.12/compat/linux/x86_64/usr/lib/python3.9/site-packages/reframe/frontend/loader.py", line 237, in load_from_file
    util.import_module_from_file(filename, force)
  File "/cvmfs/pilot.eessi-hpc.org/2021.12/compat/linux/x86_64/usr/lib/python3.9/site-packages/reframe/utility/__init__.py", line 91, in import_module_from_file
    return _do_import_module_from_file(filename, module_name)
  File "/cvmfs/pilot.eessi-hpc.org/2021.12/compat/linux/x86_64/usr/lib/python3.9/site-packages/reframe/utility/__init__.py", line 68, in _do_import_module_from_file
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/ec2-user/compatibility-layer/test/compat_layer.py", line 11, in <module>
    class RunInGentooPrefixTest(rfm.RunOnlyRegressionTest):
  File "/home/ec2-user/compatibility-layer/test/compat_layer.py", line 17, in RunInGentooPrefixTest
    if eessi_version != '2021.06':
  File "/cvmfs/pilot.eessi-hpc.org/2021.12/compat/linux/x86_64/usr/lib/python3.9/site-packages/reframe/core/meta.py", line 143, in __getitem__
    raise ReframeSyntaxError(
reframe.core.exceptions.ReframeSyntaxError: accessing a test parameter from the class body is disallowed

Loaded 0 test(s)
Generated 0 test case(s)
Filtering test cases(s) by name: 0 remaining
Filtering test cases(s) by tags: 0 remaining
Filtering test cases(s) by other attributes: 0 remaining
Final number of test cases: 0
[==========] Running 0 check(s)
[==========] Started on Tue Nov 30 07:52:29 2021

[----------] waiting for spawned checks to finish
[----------] all spawned checks have finished

[  PASSED  ] Ran 0/0 test case(s) from 0 check(s) (0 failure(s), 0 skipped)
[==========] Finished on Tue Nov 30 07:52:29 2021
Run report saved in '/home/ec2-user/.reframe/reports/run-report.json'
Log file(s) saved in '/tmp/rfm-m33a1fsh.log'

$ echo $?
0
```

**Comment by vkarak** [comment]
Hi @boegel, this is not a bug and let me explain why. When you load multiple tests and a single one is problematic, we intentionally don't want to fail the whole test loading process and not run anything. This has been in ReFrame quite some time. In those cases, we report the error and continue and ask the user to add `-v` in order to get more info and a stack trace. The fact that this particular case was failing hard in 3.6.2 was a bug, because as I said before we explicitly don't want to stop the loading process if a test is problematic.

Now what is a `ReframeSyntaxError`. When ReFrame detects a likely programming error (i.e., `AttributeError`, `NameError`, `TypeError`, `ValueError`) it tries to trace it back to the offending user source code line, in which case it will report it. If not, it will issue a stack trace assuming that it is a framework problem. With the introduction of built-ins, such as `variable`s and `parameter`s, it is not always feasible to trace this back to the user line (at the moment), because there are several validation steps that can only happen at a point later than the definition of the user variable or parameter. That's why we introduced the `ReframeSyntaxError`, to capture such programming errors and don't crash badly for user errors (as in previous version).

In summary, this is not a bug, but I could see two feature requests coming out from this issue:

- Add a `--stop-on-load-errors` option that would force reframe to stop on any error during loading a test.
- Provide source code information in `ReFrameSyntaxError`s.

**Comment by boegel** [comment]
I understand the motivation for not making this an error by default and continuing to try and run load tests from other files and run those. Having an option to treat incorrect `*.py` files as an error would definitely be welcome.

I think it would make sense to *not* exit with exit code 0 if one or more test files could not be loaded (file not found, syntax error, whatever), because that's very misleading, and it could easily lead to incorrect conclusions unless a human is actually taking a look at the generated output. So maybe loading the test module should be some kind of implicit test, which can fail?

**Comment by vkarak** [comment]
Fair enough, I will also open a feature request for that.

**Comment by vkarak** [comment]
Feature requests opened: #2306, #2307, #2308.
